test/library/library.go

```diff
@@ -32,3 +47,117 @@ func GenerateNameForTest(t *testing.T, prefix string) string {
 	// ensure no `-` at beginning or end
 	return strings.Trim(string(name), "-")
 }
+
+// WaitForAPIServerToBeOnTheSameRevision waits until all API Servers are running at the same revision.
+// The API Servers must stay on the same revision for at least waitForAPIRevisionSuccessThreshold * waitForAPIRevisionSuccessInterval
+//
+// Observations:
+//  rolling out a new version is not instant you need to account for a propagation time (~1/2 minutes)
+//  rolling out a new version can take ~10 minutes
+func WaitForAPIServerToBeOnTheSameRevision(t *testing.T, podClient corev1client.PodInterface) {
+	if err := wait.Poll(waitForAPIRevisionPollInterval, waitForAPIRevisionTimeout, mustSucceedMultipleTimes(waitForAPIRevisionSuccessThreshold, waitForAPIRevisionSuccessInterval, func() (bool, error) {
+		converged, err := areAPIServersOnTheSameRevision(podClient)
+		return converged, err
+	})); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// areAPIServersOnTheSameRevision tries to find the current revision that the API servers are running at.
+// The number of instances is calculated based on the number of running pods in a namespace.
+// This should be okay because this function is meant to be used by WaitForAPIServerToBeOnTheSameRevision which will wait at least waitForAPIRevisionSuccessThreshold * waitForAPIRevisionSuccessInterval
+// The number of pods should stabilize in that period of time.
+func areAPIServersOnTheSameRevision(podClient corev1client.PodInterface) (bool, error) {
+	revisionLabel := "revision"
+
+	// do a live list so we never get confused about what revision we are on
+	apiServerPods, err := podClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "apiserver=true"})
+	if err != nil {
+		return false, err
+	}
+
+	good, bad, progressing, err := categorizePods(apiServerPods.Items)
+	if err != nil {
+		return false, err
+	}
+	if progressing {
+		return false, nil
+	}
+
+	goodRevisions := revisions(revisionLabel, good)
+	failingRevisions := revisions(revisionLabel, bad)
+
+	if len(goodRevisions) != 1 {
+		return false, nil // api servers have not converged onto a single revision
+	}
+	revision, _ := goodRevisions.PopAny()
+	if len(revision) == 0 {
```

Comment:
Will this actually be an empty string or is it a legacy of the implementation being copied?

Reply:
oh, I think you are right - it was a copy/paste. I'll remove it.